pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java

```diff
@@ -62,57 +69,25 @@ public AggregationPlanNode(IndexSegment indexSegment, QueryContext queryContext)
   public Operator<IntermediateResultsBlock> run() {
     assert _queryContext.getAggregationFunctions() != null;
 
-    int numTotalDocs = _indexSegment.getSegmentMetadata().getTotalDocs();
-    AggregationFunction[] aggregationFunctions = _queryContext.getAggregationFunctions();
+    boolean hasFilteredPredicates = _queryContext.isHasFilteredAggregations();
 
-    FilterPlanNode filterPlanNode = new FilterPlanNode(_indexSegment, _queryContext);
-    BaseFilterOperator filterOperator = filterPlanNode.run();
+    Pair<FilterPlanNode, BaseFilterOperator> filterOperatorPair =
```

Comment:
Suggest splitting the logic of filtered aggregation and regular aggregation. Currently the logic is mixed in `buildOperators()` which is hard to read or debug.
We can keep the regular aggregation handling unchanged, and add the code path for the filtered aggregations. The metadata/dictionary/startree based operator does not apply to the filtered aggregation.

Reply:
buildOperators has no code specific to filtered aggregations. It builds a generic pipeline of operators, and buildOperatorForFilteredAggregations does the actual specific code if filtered aggregations are present.

Reply:
I am not sure if I understand -- why should star-tree be not used for a filtered aggregation? If a swim lane is able to use star tree to process the underlying predicate, shouldn't we allow that? Or is there a technical limitation to that?

Regarding the forking, I agree the code readability was an issue. I have fixed the same